// nitro@3.0.1-alpha.1/dist/presets/cloudflare/runtime/cloudflare-durable.mjs
import "#nitro-internal-pollyfills";
import { DurableObject } from "cloudflare:workers";
import wsAdapter from "crossws/adapters/cloudflare";
import { createHandler, fetchHandler } from "./_module-handler.mjs";
import { useNitroApp, useNitroHooks } from "nitro/app";
import { isPublicAssetURL } from "#nitro-internal-virtual/public-assets";
import { resolveWebsocketHooks } from "nitro/~internal/runtime/app";
import { hasWebSocket } from "#nitro-internal-virtual/feature-flags";
const DURABLE_BINDING = "$DurableObject";
const DURABLE_INSTANCE = "server";
const nitroApp = useNitroApp();
const nitroHooks = useNitroHooks();
const getDurableStub = (env) => {
	const binding = env[DURABLE_BINDING];
	if (!binding) {
		throw new Error(`Durable Object binding "${DURABLE_BINDING}" not available.`);
	}
	const id = binding.idFromName(DURABLE_INSTANCE);
	return binding.get(id);
};
const ws = hasWebSocket ? wsAdapter({
	resolve: resolveWebsocketHooks,
	instanceName: DURABLE_INSTANCE,
	bindingName: DURABLE_BINDING
}) : undefined;
export default createHandler({ fetch(request, env, context, url, ctxExt) {
	// Static assets fallback (optional binding)
	if (env.ASSETS && isPublicAssetURL(url.pathname)) {
		return env.ASSETS.fetch(request);
	}
	// Expose stub fetch to the context
	ctxExt.durableFetch = (req = request) => getDurableStub(env).fetch(req);
	// Websocket upgrade
	// https://crossws.unjs.io/adapters/cloudflare#durable-objects
	if (hasWebSocket && request.headers.get("upgrade") === "websocket") {
		return ws.handleUpgrade(request, env, context);
	}
} });
export class $DurableObject extends DurableObject {
	constructor(state, env) {
		super(state, env);
		state.waitUntil(nitroHooks.callHook("cloudflare:durable:init", this, {
			state,
			env
		}));
		if (hasWebSocket) {
			ws.handleDurableInit(this, state, env);
		}
	}
	fetch(request) {
		if (hasWebSocket && request.headers.get("upgrade") === "websocket") {
			return ws.handleDurableUpgrade(this, request);
		}
		// Main handler
		const url = new URL(request.url);
		return fetchHandler(request, this.env, this.ctx, url, nitroApp, { durable: this });
	}
	alarm() {
		this.ctx.waitUntil(nitroHooks.callHook("cloudflare:durable:alarm", this));
	}
	async webSocketMessage(client, message) {
		if (hasWebSocket) {
			return ws.handleDurableMessage(this, client, message);
		}
	}
	async webSocketClose(client, code, reason, wasClean) {
		if (hasWebSocket) {
			return ws.handleDurableClose(this, client, code, reason, wasClean);
		}
	}
}
