// nitro@3.0.1-alpha.1/dist/builder.mjs
import "./_libs/c12.mjs";
import "./_libs/gen-mapping.mjs";
import "./_libs/magic-string.mjs";
import "./_libs/acorn.mjs";
import "./_libs/confbox.mjs";
import "./_libs/local-pkg.mjs";
import "./_libs/js-tokens.mjs";
import "./_libs/strip-literal.mjs";
import "./_libs/unimport.mjs";
import "./_libs/picomatch.mjs";
import "./_libs/fdir.mjs";
import "./_libs/tinyglobby.mjs";
import "./_libs/compatx.mjs";
import "./_libs/klona.mjs";
import "./_libs/std-env.mjs";
import { a as createNitro, c as loadOptions, i as build, n as prepare, o as listTasks, r as copyPublicAssets, s as runTask, t as prerender } from "./_chunks/B-D1JOIz.mjs";
import "./_libs/escape-string-regexp.mjs";
import "./_libs/tsconfck.mjs";
import "./_libs/dot-prop.mjs";
import "./_chunks/C7CbzoI1.mjs";
import { n as writeTypes } from "./_chunks/ANM1K1bE.mjs";
import "./_libs/rou3.mjs";
import "./_libs/mime.mjs";
import "./_libs/pathe.mjs";
import "./_libs/untyped.mjs";
import "./_libs/knitwork.mjs";
import { t as getBuildInfo } from "./_build/common.mjs";
import "./_libs/httpxy.mjs";
import { n as createDevServer } from "./_dev.mjs";
import "./_libs/chokidar.mjs";
import "./_libs/ultrahtml.mjs";

export { build, copyPublicAssets, createDevServer, createNitro, getBuildInfo, listTasks, loadOptions, prepare, prerender, runTask, writeTypes };
